File: vm.py
import math
from opcodes import OpCode


# Built-in functions callable from pseudocode
_BUILTINS = {
    "length": lambda args: len(args[0]) if isinstance(args[0], list) else 1,
    "floor": lambda args: math.floor(args[0]),
    "ceil": lambda args: math.ceil(args[0]),
    "abs": lambda args: abs(args[0]),
    "sqrt": lambda args: math.sqrt(args[0]),
    "min": lambda args: min(args),
    "max": lambda args: max(args),
}


class VM:
    """Virtual Machine to execute generated opcodes"""

    def __init__(self, instructions, function_table=None):
        self.instructions = instructions
        self.pc = 0
        self.variables = {}
        self.last_cmp = False
        self.return_value = None
        self.call_stack = []  # (return_pc, saved_vars, result_target)
        self.function_table = function_table or {}
        self.inputs = self._detect_inputs()

    def _detect_inputs(self):
        """Detect variables that are read before being written (i.e., inputs)."""
        written = set()
        inputs = {}

        for instr in self.instructions:
            opcode = instr.opcode
            operands = instr.operands

            if opcode == OpCode.ASN:
                value = str(operands[1])
                if (
                    value.isalpha()
                    and value not in written
                    and not value.startswith("t")
                ):
                    inputs[value] = None

                target = str(operands[0])
                written.add(target.split("[")[0])

            elif opcode == OpCode.AOP:
                # Unary (3 operands) or binary (4 operands)
                read_ops = (
                    [operands[1]] if len(operands) == 3 else [operands[1], operands[2]]
                )
                for operand in read_ops:
                    op = str(operand)
                    if op.isalpha() and op not in written and not op.startswith("t"):
                        inputs[op] = None
                written.add(str(operands[-1]))

            elif opcode == OpCode.COM:
                for operand in [operands[1], operands[2]]:
                    op = str(operand)
                    if op.isalpha() and op not in written and not op.startswith("t"):
                        inputs[op] = None
                written.add(str(operands[3]))

            elif opcode == OpCode.IDX:
                for operand in [operands[0], operands[1]]:
                    op = str(operand)
                    if op.isalpha() and op not in written and not op.startswith("t"):
                        inputs[op] = None
                written.add(str(operands[2]))

        return inputs

    def get_value(self, operand):
        """Resolve an operand to its runtime value."""
        operand = str(operand)
        if operand == "true":
            return True
        if operand == "false":
            return False
        if operand in ("NIL", "null"):
            return None
        if operand == "infinity":
            return float("inf")
        # Strip string literal quotes
        if (operand.startswith('"') and operand.endswith('"')) or (
            operand.startswith("'") and operand.endswith("'")
        ):
            return operand[1:-1]
        try:
            return int(operand)
        except (ValueError, TypeError):
            return self.variables.get(operand, 0)

    def set_indexed(self, target, value):
        """Write to a plain variable or an array element (A[i])."""
        if "[" in target:
            arr_name = target.split("[")[0]
            index_str = target.split("[")[1].rstrip("]")
            index = self.get_value(index_str)
            if arr_name not in self.variables:
                self.variables[arr_name] = []
            while len(self.variables[arr_name]) <= index:
                self.variables[arr_name].append(0)
            self.variables[arr_name][index] = value
        else:
            self.variables[target] = value

    def execute(self, instr):
        """Execute one instruction; return PC delta (0 = advance by 1)."""
        opcode = instr.opcode
        operands = instr.operands

        if opcode == OpCode.ASN:
            target, value = operands[0], self.get_value(operands[1])
            self.set_indexed(target, value)

        elif opcode == OpCode.AOP:
            op = operands[0]
            if len(operands) == 3:
                # Unary operation (currently only "not")
                val = self.get_value(operands[1])
                result = str(operands[2])
                if op == "not":
                    result_val = not bool(val)
                    self.variables[result] = result_val
                    self.last_cmp = result_val  # keep last_cmp in sync for control flow
                else:
                    raise ValueError(f"Unknown unary op: {op}")
            else:
                # Binary arithmetic
                op, left, right, result = operands
                left_val = self.get_value(left)
                right_val = self.get_value(right)

                if op == "+":
                    self.variables[result] = left_val + right_val
                elif op == "-":
                    self.variables[result] = left_val - right_val
                elif op == "*":
                    self.variables[result] = left_val * right_val
                elif op == "/":
                    self.variables[result] = left_val // right_val
                elif op == "mod":
                    self.variables[result] = left_val % right_val
                else:
                    raise ValueError(f"Unknown arithmetic op: {op}")

        elif opcode == OpCode.COM:
            op, left, right, result = operands
            left_val = self.get_value(left)
            right_val = self.get_value(right)

            if op == "<":
                cmp_result = left_val < right_val
            elif op == ">":
                cmp_result = left_val > right_val
            elif op == "<=":
                cmp_result = left_val <= right_val
            elif op == ">=":
                cmp_result = left_val >= right_val
            elif op == "=":
                cmp_result = left_val == right_val
            elif op == "!=":
                cmp_result = left_val != right_val
            elif op == "and":
                cmp_result = bool(left_val) and bool(right_val)
            elif op == "or":
                cmp_result = bool(left_val) or bool(right_val)
            else:
                cmp_result = False

            self.variables[result] = cmp_result
            self.last_cmp = cmp_result

        elif opcode == OpCode.IDX:
            array_name, index, result = operands
            index_val = self.get_value(index)
            arr = self.variables.get(array_name, [])
            self.variables[result] = arr[index_val] if index_val < len(arr) else 0

        elif opcode == OpCode.CAL:
            func_name = str(operands[0])
            arg_operands = operands[1:-1]
            result_target = str(operands[-1])
            arg_values = [self.get_value(a) for a in arg_operands]

            if func_name in _BUILTINS:
                result = _BUILTINS[func_name](arg_values)
                self.variables[result_target] = result
            else:
                if func_name not in self.function_table:
                    raise RuntimeError(f"Unknown function: {func_name!r}")
                start_idx, param_names = self.function_table[func_name]
                # Push caller frame
                self.call_stack.append(
                    (self.pc + 1, dict(self.variables), result_target)
                )
                # New scope: only function parameters
                self.variables = {
                    name: val for name, val in zip(param_names, arg_values)
                }
                return start_idx - self.pc - 1

        elif opcode == OpCode.PRT:
            print(self.get_value(operands[0]))

        elif opcode == OpCode.SKP:
            n = int(operands[0])
            if not self.last_cmp:
                return n

        elif opcode == OpCode.SKPT:
            n = int(operands[0])
            if self.last_cmp:
                return n

        elif opcode == OpCode.JMP:
            target = int(operands[0])
            return target - self.pc - 1

        elif opcode == OpCode.RET:
            ret_val = self.get_value(operands[0]) if operands else None

            if self.call_stack:
                return_pc, saved_vars, result_target = self.call_stack.pop()
                self.variables = saved_vars
                if result_target and ret_val is not None:
                    self.variables[result_target] = ret_val
                return return_pc - self.pc - 1
            else:
                self.return_value = ret_val
                return len(self.instructions)  # halt

        return 0

    def run(self, **initial_vars):
        """Run program with initial variable bindings."""
        self.variables.update(initial_vars)
        self.pc = 0
        self.last_cmp = False
        self.return_value = None

        while self.pc < len(self.instructions):
            pc_delta = self.execute(self.instructions[self.pc])
            self.pc += 1 + pc_delta

        return self.return_value

File: go/vm.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

type CallFrame struct {
	returnPC     int
	savedVars    map[string]Value
	resultTarget string
}

type VM struct {
	instrs    []Instruction
	funcTable map[string]FuncEntry
	pc        int
	vars      map[string]Value
	lastCmp   bool
	retVal    *Value
	callStack []CallFrame
}

func NewVM(instrs []Instruction, funcTable map[string]FuncEntry) *VM {
	return &VM{
		instrs:    instrs,
		funcTable: funcTable,
		vars:      make(map[string]Value),
	}
}

func (vm *VM) getVal(s string) Value {
	if (strings.HasPrefix(s, `"`) && strings.HasSuffix(s, `"`)) ||
		(strings.HasPrefix(s, `'`) && strings.HasSuffix(s, `'`)) {
		return IntVal(0)
	}
	if n, err := strconv.Atoi(s); err == nil {
		return IntVal(n)
	}
	if v, ok := vm.vars[s]; ok {
		return v
	}
	return IntVal(0)
}

func (vm *VM) setIndexed(target string, val Value) {
	if !strings.Contains(target, "[") {
		vm.vars[target] = val
		return
	}
	bracket := strings.Index(target, "[")
	arrName := target[:bracket]
	idxStr := target[bracket+1 : len(target)-1]
	idx := toInt(vm.getVal(idxStr))

	arr, ok := vm.vars[arrName]
	if !ok || arr.Kind != TypeArray {
		arr = ArrayVal(nil)
	}
	for len(arr.ArrVal) <= idx {
		arr.ArrVal = append(arr.ArrVal, IntVal(0))
	}
	arr.ArrVal[idx] = val
	vm.vars[arrName] = arr
}

func (vm *VM) execOne(instr Instruction) (int, error) {
	ops := instr.Operands

	switch instr.Opcode {
	case ASN:
		if len(ops) > 2 {
			elems := make([]Value, len(ops)-1)
			for i, e := range ops[1:] {
				elems[i] = vm.getVal(e)
			}
			vm.setIndexed(ops[0], ArrayVal(elems))
		} else {
			vm.setIndexed(ops[0], vm.getVal(ops[1]))
		}

	case AOP:
		op := ops[0]
		if len(ops) == 3 {
			result := !toBool(vm.getVal(ops[1]))
			vm.vars[ops[2]] = boolToVal(result)
			vm.lastCmp = result
		} else {
			l, r := toInt(vm.getVal(ops[1])), toInt(vm.getVal(ops[2]))
			var res int
			switch op {
			case "+":
				res = l + r
			case "-":
				res = l - r
			case "*":
				res = l * r
			case "/":
				if r == 0 {
					return 0, fmt.Errorf("division by zero")
				}
				res = l / r
			case "mod":
				if r == 0 {
					return 0, fmt.Errorf("modulo by zero")
				}
				res = l % r
			default:
				return 0, fmt.Errorf("unknown AOP: %s", op)
			}
			vm.vars[ops[3]] = IntVal(res)
		}

	case COM:
		op, leftOp, rightOp, result := ops[0], ops[1], ops[2], ops[3]
		lv := toInt(vm.getVal(leftOp))
		rv := toInt(vm.getVal(rightOp))
		var cmp bool
		switch op {
		case "<":
			cmp = lv < rv
		case ">":
			cmp = lv > rv
		case "<=":
			cmp = lv <= rv
		case ">=":
			cmp = lv >= rv
		case "=":
			cmp = lv == rv
		case "!=":
			cmp = lv != rv
		case "and":
			cmp = toBool(vm.getVal(leftOp)) && toBool(vm.getVal(rightOp))
		case "or":
			cmp = toBool(vm.getVal(leftOp)) || toBool(vm.getVal(rightOp))
		default:
			cmp = false
		}
		vm.vars[result] = boolToVal(cmp)
		vm.lastCmp = cmp

	case IDX:
		arrName, idxOp, result := ops[0], ops[1], ops[2]
		arr := vm.vars[arrName]
		idx := toInt(vm.getVal(idxOp))
		if arr.Kind == TypeArray && idx >= 0 && idx < len(arr.ArrVal) {
			vm.vars[result] = arr.ArrVal[idx]
		} else {
			vm.vars[result] = IntVal(0)
		}

	case CAL:
		funcName := ops[0]
		argOps := ops[1 : len(ops)-1]
		resultTarget := ops[len(ops)-1]
		argVals := make([]Value, len(argOps))
		for i, a := range argOps {
			argVals[i] = vm.getVal(a)
		}
		if builtin, ok := builtins[funcName]; ok {
			res, err := builtin(argVals)
			if err != nil {
				return 0, err
			}
			vm.vars[resultTarget] = res
			break
		}
		entry, ok := vm.funcTable[funcName]
		if !ok {
			return 0, fmt.Errorf("unknown function: %q", funcName)
		}
		vm.callStack = append(vm.callStack, CallFrame{vm.pc + 1, deepCopyVars(vm.vars), resultTarget})
		newVars := make(map[string]Value, len(entry.ParamNames))
		for i, name := range entry.ParamNames {
			if i < len(argVals) {
				newVars[name] = argVals[i]
			}
		}
		vm.vars = newVars
		return entry.StartIdx - vm.pc - 1, nil

	case PRT:
		fmt.Println(vm.getVal(ops[0]).Format())

	case SKP:
		n, _ := strconv.Atoi(ops[0])
		if !vm.lastCmp {
			return n, nil
		}

	case SKPT:
		n, _ := strconv.Atoi(ops[0])
		if vm.lastCmp {
			return n, nil
		}

	case JMP:
		target, _ := strconv.Atoi(ops[0])
		return target - vm.pc - 1, nil

	case RET:
		var rv Value
		if len(ops) > 0 {
			rv = vm.getVal(ops[0])
		} else {
			rv = Nil
		}
		if len(vm.callStack) > 0 {
			frame := vm.callStack[len(vm.callStack)-1]
			vm.callStack = vm.callStack[:len(vm.callStack)-1]
			vm.vars = frame.savedVars
			if frame.resultTarget != "" {
				vm.vars[frame.resultTarget] = rv
			}
			return frame.returnPC - vm.pc - 1, nil
		}
		vm.retVal = &rv
		return len(vm.instrs), nil
	}
	return 0, nil
}

var builtins = map[string]func([]Value) (Value, error){
	"length": func(args []Value) (Value, error) {
		if args[0].Kind == TypeArray {
			return IntVal(len(args[0].ArrVal)), nil
		}
		return IntVal(1), nil
	},
	"floor": func(args []Value) (Value, error) {
		return IntVal(int(math.Floor(float64(toInt(args[0]))))), nil
	},
	"ceil": func(args []Value) (Value, error) {
		return IntVal(int(math.Ceil(float64(toInt(args[0]))))), nil
	},
	"abs": func(args []Value) (Value, error) {
		n := toInt(args[0])
		if n < 0 {
			n = -n
		}
		return IntVal(n), nil
	},
	"sqrt": func(args []Value) (Value, error) {
		return IntVal(int(math.Sqrt(float64(toInt(args[0]))))), nil
	},
	"min": func(args []Value) (Value, error) {
		m := toInt(args[0])
		for _, a := range args[1:] {
			if v := toInt(a); v < m {
				m = v
			}
		}
		return IntVal(m), nil
	},
	"max": func(args []Value) (Value, error) {
		m := toInt(args[0])
		for _, a := range args[1:] {
			if v := toInt(a); v > m {
				m = v
			}
		}
		return IntVal(m), nil
	},
}

func (vm *VM) Run(initial map[string]Value) (Value, error) {
	for k, v := range initial {
		vm.vars[k] = v
	}
	vm.pc = 0
	vm.lastCmp = false
	vm.retVal = nil

	for vm.pc < len(vm.instrs) {
		delta, err := vm.execOne(vm.instrs[vm.pc])
		if err != nil {
			return Nil, err
		}
		vm.pc += 1 + delta
	}

	if vm.retVal != nil {
		return *vm.retVal, nil
	}
	return Nil, nil
}
